qos/evm/evm.go

```diff
@@ -25,23 +25,46 @@ var _ gateway.QoSService = &QoS{}
 //   - Response building
 //   - Endpoint validation and selection
 type QoS struct {
+	logger polylog.Logger
 	*EndpointStore
 	*ServiceState
-	logger polylog.Logger
 }
 
 // ParseHTTPRequest builds a request context from an HTTP request.
 // Returns (context, false) if POST request is not valid JSON-RPC.
 // Implements gateway.QoSService interface.
 func (qos *QoS) ParseHTTPRequest(_ context.Context, req *http.Request) (gateway.RequestQoSContext, bool) {
+	logger := qos.logger.With(
+		"qos", "EVM",
+		"method", "ParseHTTPRequest",
+	)
+
+	// TODO_TECHDEBT(@adshmh): Simplify the qos package by refactoring gateway.QoSContextBuilder.
+	// Proposed change: Create a new ServiceRequest type containing raw payload data ([]byte)
+	// Benefits: Decouples the qos package from HTTP-specific error handling.
 	body, err := io.ReadAll(req.Body)
 	if err != nil {
-		return requestContextFromInternalError(err), false
+		logger.Warn().Err(err).Msg("HTTP request body read failed - returning generic error response.")
+
+		return requestContextFromInternalError(
+			qos.logger,
+			err,
+			qosobservations.EVMRequestValidationErrorKind_EVM_REQUEST_VALIDATION_ERROR_KIND_HTTP_BODY_READ_FAILURE,
+		), false
 	}
 
+	// TODO_MVP(@adshmh): Add a JSON-RPC request validator to reject invalid/unsupported method calls early in request flow.
 	var jsonrpcReq jsonrpc.Request
 	if err := json.Unmarshal(body, &jsonrpcReq); err != nil {
-		return requestContextFromUserError(err), false
+		logger.With(
+			"request_preview", string(body[:min(1000, len(body))]), // truncate body to first 1000 bytes for logging.
```

Comment:
TIL about `min`.

Reply:
`min` and `max` were introduced in go 1.21.